sdk/src/contexts/solana/relayer.ts

```diff
@@ -0,0 +1,186 @@
+  }
+
+  async calculateRelayerFee(
+    targetChain: ChainId,
+    mint: PublicKey,
+    decimals: number,
+  ): Promise<bigint> {
+    const [{ fee }, { swapRate }, { relayerFeePrecision, swapRatePrecision }] =
+      await Promise.all([
+        this.getRelayerFee(targetChain),
+        this.getRegisteredToken(mint),
+        this.getRedeemerConfig(),
+      ]);
+    const relayerFee = TEN.pow(new BN(decimals))
+      .mul(fee)
+      .mul(new BN(swapRatePrecision))
+      .div(new BN(relayerFeePrecision).mul(swapRate));
+
+    return BigInt(relayerFee.toString());
+  }
+
+  async calculateMaxSwapAmountIn(
+    mint: PublicKey,
+    decimals: number,
+  ): Promise<bigint> {
+    const [
+      { swapRate, maxNativeSwapAmount },
+      { swapRate: solSwapRate },
+      { swapRatePrecision },
+    ] = await Promise.all([
+      this.getRegisteredToken(mint),
+      this.getRegisteredToken(NATIVE_MINT),
+      this.getRedeemerConfig(),
+    ]);
+    const swapRatePrecisionBN = new BN(swapRatePrecision);
+    const nativeSwapRate = this.calculateNativeSwapRate(
+      swapRatePrecisionBN,
+      solSwapRate,
+      swapRate,
+    );
+    const maxSwapAmountIn =
+      decimals > SOL_DECIMALS
+        ? maxNativeSwapAmount
+            .mul(nativeSwapRate)
+            .mul(TEN.pow(new BN(decimals - SOL_DECIMALS)))
+            .div(swapRatePrecisionBN)
+        : maxNativeSwapAmount
+            .mul(nativeSwapRate)
+            .div(
+              TEN.pow(new BN(SOL_DECIMALS - decimals)).mul(swapRatePrecisionBN),
+            );
+
+    return BigInt(maxSwapAmountIn.toString());
+  }
+
+  async calculateNativeSwapAmountOut(
+    mint: PublicKey,
+    toNativeAmount: bigint,
+    decimals: number,
+  ): Promise<bigint> {
+    if (toNativeAmount === 0n) {
+      return 0n;
+    }
+    const [{ swapRate }, { swapRate: solSwapRate }, { swapRatePrecision }] =
+      await Promise.all([
+        this.getRegisteredToken(mint),
+        this.getRegisteredToken(NATIVE_MINT),
+        this.getRedeemerConfig(),
+      ]);
+    const swapRatePrecisionBN = new BN(swapRatePrecision);
+    const nativeSwapRate = this.calculateNativeSwapRate(
+      swapRatePrecisionBN,
+      solSwapRate,
+      swapRate,
+    );
+    const swapAmountOut =
+      decimals > SOL_DECIMALS
+        ? swapRatePrecisionBN
+            .mul(new BN(toNativeAmount.toString()))
+            .div(nativeSwapRate.mul(TEN.pow(new BN(decimals - SOL_DECIMALS))))
+        : swapRatePrecisionBN
+            .mul(new BN(toNativeAmount.toString()))
+            .mul(TEN.pow(new BN(SOL_DECIMALS - decimals)))
+            .div(nativeSwapRate);
+
+    return BigInt(swapAmountOut.toString());
+  }
+
+  async fetchSwapEvent(signature: string): Promise<SwapEvent | null> {
+    const transaction = await this.connection.getParsedTransaction(signature);
+    if (transaction) {
+      const logMessages = transaction.meta?.logMessages || [];
+      for (const msg of logMessages) {
+        const matches =
```

Comment:
Is there any other way to retrieve this data or is using a regex the only "elegant" way?

Reply:
Unfortunately this is how we have to do it on Solana. The contract calls `msg!` which is just a macro that logs data in the transaction.